cv32e40s/tests/programs/custom/debug_test2/debug_test2.c

```diff
@@ -0,0 +1,3814 @@
+}
+
+// -----------------------------------------------------------------------------
+
+// Illegal csr in dmode debug code
+void __attribute__((naked)) illegal_csr_in_dmode_dbg(void) {
+  __asm__ volatile ( R"(
+    csrrw s10, 0xeaf, s10
+  )");
+}
+
+// -----------------------------------------------------------------------------
+
+// Ecall in dmode test entry
+uint32_t ecall_in_dmode(uint32_t index, uint8_t report_name) {
+  volatile uint8_t test_fail = 0;
+  volatile debug_req_control_t debug_req_ctrl;
+
+  SET_FUNC_INFO
+
+  if (report_name) {
+    cvprintf(V_LOW, "\"%s\"", name);
+    return 0;
+  }
+
+  *g_debug_status   = 0;
+  *g_debug_test_num = 6;
+
+  debug_req_ctrl = (debug_req_control_t) {
+    .fields.value            = 1,
+    .fields.pulse_mode       = 1,
+    .fields.rand_pulse_width = 0,
+    .fields.pulse_width      = 0x1fff,
+    .fields.rand_start_delay = 0,
+    .fields.start_delay      = 200
+  };
+
+  DEBUG_REQ_CONTROL_REG = debug_req_ctrl.raw;
+
+  while (*g_debug_status == 0) {
+    ;;
+  }
+
+  test_fail += (*g_debug_status == 1) ? 0 : 1;
+
+  if (test_fail) {
+    cvprintf(V_LOW, "\nTest: \"%s\" FAIL!\n", name);
+    return index + 1;
+  }
+  cvprintf(V_MEDIUM, "\nTest: \"%s\" OK!\n", name);
+  return 0;
+}
+
+// -----------------------------------------------------------------------------
+
+// Debug mode ecall
+void ecall_in_dmode_dbg(void) {
+  __asm__ volatile ( R"(
+    ecall
+  )");
+}
+
+// -----------------------------------------------------------------------------
+
+// Mret in dmode test entry
+uint32_t mret_in_dmode(uint32_t index, uint8_t report_name) {
+  volatile uint8_t test_fail = 0;
+  volatile debug_req_control_t debug_req_ctrl;
+
+  SET_FUNC_INFO
+
+  if (report_name) {
+    cvprintf(V_LOW, "\"%s\"", name);
+    return 0;
+  }
+
+  *g_debug_status   = 0;
+  *g_debug_test_num = 7;
+
+  debug_req_ctrl = (debug_req_control_t) {
+    .fields.value            = 1,
+    .fields.pulse_mode       = 1,
+    .fields.rand_pulse_width = 0,
+    .fields.pulse_width      = 0x1fff,
+    .fields.rand_start_delay = 0,
+    .fields.start_delay      = 200
+  };
+
+  DEBUG_REQ_CONTROL_REG = debug_req_ctrl.raw;
+
+  while (*g_debug_status == 0) {
+    ;;
+  }
+
+  test_fail += (*g_debug_status == 1) ? 0 : 1;
+
+  if (test_fail) {
+    cvprintf(V_LOW, "\nTest: \"%s\" FAIL!\n", name);
+    return index + 1;
+  }
+  cvprintf(V_MEDIUM, "\nTest: \"%s\" OK!\n", name);
+  return 0;
+}
+
+// -----------------------------------------------------------------------------
+
+// Mret in debug mode, debug portion
+// This does technically not need to be in a function
+// but for test structure and readability keeps the
+// structure used for other tests.
+void mret_in_dmode_dbg(void) {
+  __asm__ volatile ( R"(
+    mret
+  )");
+}
+
+// -----------------------------------------------------------------------------
+
+// Exception to dmode test entry
+uint32_t exception_enters_debug_mode(uint32_t index, uint8_t report_name) {
```

Comment:
How does this test work? The M-mode part illegally accesses a CSR, so the u_sw... handler sees the test number and calls c.ebreak, and then... does it rely on an earlier test setting "ebreakm" or what happens next? Or is there an exception trigger previously set up?

I am missing something.

Should it clear the "expecting exception" variable?

Reply:
This test warrants a comment - the test goal here is to check that debug mode can be entered from an exception handler - u_sw_... calls ebreak after trapping entering debug mode, and debug code (```exception_enters_debug_mode_dbg```) sets the g_debug_status variable appropriately to ensure that we reached the correct portion of code.